### ISSUE TITLE
"dryRun" has inconsistent/unclear semantics in plugins

### ISSUE THREAD

**Issue by rbellamy** [intuit/auto#1279]
## Description

In the documentation, `--dry-run` is repeatedly described as, essentially, "Report what command will do but do not actually do anything."

This is almost universally incorrect when it comes to the packaging plugins.

### Documentation

When looking at the documentation at the areas where the `dryRun` option is available to the hooks, it appears it is only available during `makeRelease`. This turns out to NOT be true, as `auto.options.dryRun` will exist, regardless of the hook, if the command-line switch is included in the command-line.

It is unclear what the guidance is for how `--dry-run` plays out with respect to versions in package management files (`gemspec`, `gradle.properties`, `package.json`, `pom.xml`).

### Plugins

Undoubtedly, there are plugins that make changes to the packaged versions and push those changes to the git repository, regardless of whether `--dry-run` is used or not. From what I can tell, the only plugin that attempts to honor that switch is the `npm` plugin:

```console
2020-06-07 10:44:20 (⎈ |eks_eks-dev-cluster:default)
rbellamy@eanna i ~/Development/Terradatum/auto/plugins remove-maven-release-plugin-requirement % grep -R "dryRun" *   
npm/src/index.ts:        "dryRun" in auto.options && auto.options.dryRun
npm/src/index.ts:      if (!options.dryRun && isIndependent) {
npm/dist/index.js:                "dryRun" in auto.options && auto.options.dryRun
npm/dist/index.js:                if (!options.dryRun && isIndependent) {
```

## `dryRun` vs ???

How **should** a plugin work with `dryRun` semantics? Should it make any changes at all to the file system, or just report those changes it would make if it were to make them? This then begs the question of whether or not you can easily show all the changes you would be making or the artifacts you would be building without actually making some changes.

For instance: "Show me the artifacts I would create with the new version." That's a very difficult task if you aren't able to make changes to the package management files that are used to drive the artifact's version.

**Comment by hipstersmoothie** [comment]
Good callout here.

> How should a plugin work with dryRun semantics?

In most cases the dryRun flag means the main hooks will not be called and `auto` will really just try to guess what it will do.

- `changelog` - only prints new release notes (no title or plugins show what they would do)
- `release` - will call it's hook with dry run. Really the only hook that should be called during a dry run. NPM is the only plugin that taps this hook to make sub-package releases for monorepo
- `shipit` -  kind of guesses what the version auto will produce is. assumes 1 package situation. doesn't call any of the publishing hooks

When `auto` was npm only there really wasn't the worry of `what build artifacts will be produced` because it was assumed that your package was already built.

> Should it make any changes at all to the file system, or just report those changes it would make if it were to make them? 

From my npm perspective it should not actually make any changes to the repo it just reports what it would do.

> This then begs the question of whether or not you can easily show all the changes you would be making or the artifacts you would be building without actually making some changes.

Very true. `auto`'s dry runs involve lots of guessing done by core and not the plugins.

---

As I see more package plugins come to auto I'm realizing there are two package publishing scenarios that are common

1. build before publish
2. build during publish

`auto` until this point has been primarily serving `1` but `2` is becoming more prominent.

I'm open to calling the hooks in different way. for v10 I think taking a look at the hooks system to see what we can improve. You point of view helps a lot! Maybe we could move the dry run checking more to the hooks to get a better experience.

**Comment by rbellamy** [comment]
What caught my attention was my work on the `maven` plugin - and the fact that during `--dry-run` the `afterShipIt` hook is called, and for both `maven` and `gradle` this hook is used to increment a SNAPSHOT version and then push the change to the central repo.

In other words, during a `--dry-run`, when using `gradle` or `maven`, you get changes to disk AND pushes to your central repository.

I would elaborate on your two scenarios, and for each include "inspection", "computation", "changes made to state before publish" and then finally, "pushing changes to state".

1. inspection - what is the version, is it a monorepo, are there properties and options which will affect how I read that version (e.g. for `gradle` and `maven` - is it a SNAPSHOT release?).
2. computation - compute the new version, compute the changelog, etc.
3. changes made to state before publish - updating the package management files to reflect the newly computed version, committing those changes to git, tagging those commits, etc.
4. build
5. push changes to state - publish, push to central repo, github release, etc.

One thing that I'm still challenged by is the order and likelihood of a hook being run. For instance, I was surprised to learn that `--dry-run` does NOT run the `version` hook.

**Comment by hipstersmoothie** [comment]
Making a graph to hopefully clear it up. Just found a bug too in doing it

**Comment by hipstersmoothie** [comment]
![Untitled-2020-04-22-1020](https://user-images.githubusercontent.com/1192452/83996362-ccdfa080-a910-11ea-9afc-c15ee5ea01d0.png)

**Comment by rbellamy** [comment]
That's super helpful!

**Comment by adierkens** [comment]
:rocket: Issue was released in [`v10.0.0-next.1`](https://github.com/intuit/auto/releases/tag/v10.0.0-next.1) :rocket:

**Comment by adierkens** [comment]
:rocket: Issue was released in [`v10.0.0`](https://github.com/intuit/auto/releases/tag/v10.0.0) :rocket:
